pkg/generate/crossplane/crossplane.go

```diff
@@ -14,157 +14,47 @@
-type templateAPIVars struct {
-	templateset.MetaVars
-	EnumDefs []*ackmodel.EnumDef
-	TypeDefs []*ackmodel.TypeDef
-	Imports  map[string]string
-}
+type GenerationOption func(*Generation)
 
-// templateCRDVars contains template variables for the template that outputs Go
-// code for a single top-level resource's API definition
-type templateCRDVars struct {
-	templateset.MetaVars
-	CRD *ackmodel.CRD
+func WithInitializer(i Initializer) GenerationOption {
+	return func(g *Generation) {
+		g.Initializers = append(g.Initializers, i)
+	}
 }
 
-// Crossplane returns a pointer to a TemplateSet containing all the templates for
-// generating Crossplane API types and controller code for an AWS service API
-func Crossplane(
-	g *generate.Generator,
-	templateBasePath string,
-) (*templateset.TemplateSet, error) {
-	enumDefs, err := g.GetEnumDefs()
-	if err != nil {
-		return nil, err
-	}
-	typeDefs, typeImports, err := g.GetTypeDefs()
-	if err != nil {
-		return nil, err
-	}
-	crds, err := g.GetCRDs()
-	if err != nil {
-		return nil, err
+func NewGeneration(g *generate.Generator, ts *templateset.TemplateSet, opts ...GenerationOption) *Generation {
+	generation := &Generation{Generator: g, TemplateSet: ts}
+	for _, f := range opts {
+		f(generation)
 	}
+	return generation
+}
 
-	ts := templateset.New(
-		templateBasePath,
-		includePaths,
-		copyPaths,
-		funcMap,
-	)
+type Generation struct {
+	TemplateSet *templateset.TemplateSet
+	Generator   *generate.Generator
 
-	metaVars := g.MetaVars()
+	Initializers []Initializer
+}
 
-	// First add all the CRDs and API types
-	apiVars := &templateAPIVars{
-		metaVars,
-		enumDefs,
-		typeDefs,
-		typeImports,
-	}
-	for _, path := range apisTemplatePaths {
-		outPath := "zz_" + strings.TrimSuffix(filepath.Base(path), ".tpl")
-		if err = ts.Add(outPath, path, apiVars); err != nil {
+// Run starts the generation flow.
+func (g *Generation) Run() (map[string]*bytes.Buffer, error) {
+	meta := g.Generator.MetaVars()
+	for _, init := range g.Initializers {
+		if err := init(g.Generator, g.TemplateSet, meta); err != nil {
 			return nil, err
 		}
 	}
-	for _, crd := range crds {
-		crdFileName := filepath.Join(
-			"apis", metaVars.ServiceIDClean, metaVars.APIVersion,
-			"zz_"+strcase.ToSnake(crd.Kind)+".go",
-		)
-		crdVars := &templateCRDVars{
-			metaVars,
-			crd,
-		}
-		if err = ts.Add(crdFileName, "apis/crd.go.tpl", crdVars); err != nil {
-			return nil, err
-		}
-	}
-
-	// Next add the controller package for each CRD
-	targets := []string{
-		"controller.go.tpl",
-		"conversions.go.tpl",
-		"hooks.go.tpl",
-	}
-	for _, crd := range crds {
-		for _, target := range targets {
-			outPath := filepath.Join(
-				"pkg", "controller", metaVars.ServiceIDClean, crd.Names.Snake,
-				"zz_"+target,
-			)
-			crdVars := &templateCRDVars{
-				metaVars,
-				crd,
-			}
-			if err = ts.Add(outPath, "pkg/"+target, crdVars); err != nil {
-				return nil, err
-			}
-		}
```

Comment:
The previous code nicely laid out the generation of templates/files including code comments about what was happening at the different stages. That is now hidden behind the Initializer things. I'm wondering what benefit the Initializers have over the previous code? I understand that the `hooks.go` file should not have had a `zz_` prefix on it, but couldn't that have just been generated once instead of in this CRD loop?

Reply:
Similar to my [comment](https://github.com/aws/aws-controllers-k8s/pull/559#discussion_r541111665) above, I separated the stages in the code so that we can test them separately.

> I understand that the hooks.go file should not have had a zz_ prefix on it, but couldn't that have just been generated once instead of in this CRD loop?

Yes, that's a viable approach. Since `hooks.go` is kind of an exception in the whole pipeline, I wanted to deal with it at a higher level so that the actual generation function can be re-used in different places without being worried about what gets eliminated. Essentially, it's to keep the initializer functions do only one thing (initialize the templates) and removal of `hooks.go` didn't feel like belonging to _templating_ part of the logic.